gui-agent/vmside.c

```diff
@@ -2188,6 +2188,12 @@ static void handle_sigterm()
     exit(0);
 }
 
+static void handle_sigchld()
+{
+    fprintf(stderr, "Xorg died unexpectedly, exiting!\n");
+    exit(1);
+}
+
```

Comment:
`fprintf` and `exit` are not async-signal-safe, so one cannot call them from a signal handler that could interrupt another async-signal-unsafe function.  If one uses `sigprocmask` or `pthread_sigmask` to block SIGCHLD except during `ppoll()`, then the problem goes away, as `ppoll()` is async-signal-safe.  For this to work, SIGCHLD must be blocked before the handler is registered.

Reply:
I can't see the `ppoll` call you're talking about.  I do see one to `select`, though. Considering it's in `gui-common/txrx-vchan.c`, though, I'd gather that doing that there would be quite painful. Perhaps it'd be better to set a flag that the main loop would then check? Not sure whether we'd get stuck in `wait_for_vchan_or_argfd`, though. If we'll exit that function with a dead Xorg, I could check the flag in the main loop quite comfortably, but otherwise this seems like a painful thing.

Reply:
You can use `pselect()` or the self-pipe trick.